UefiPayloadPkg/ApuConfigurationUi/ApuConfigurationUi.c

```diff
@@ -0,0 +1,176 @@
+/** @file
+  Dasharo APU Configuration UI for Setup Front Page.
+
+  Copyright (c) 2024, 3mdeb All rights reserved.<BR>
+
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <Guid/ApuConfigurationGuid.h>
+#include <Library/UefiDriverEntryPoint.h>
+#include <Library/UefiBootServicesTableLib.h>
+#include <Library/UefiRuntimeServicesTableLib.h>
+#include <Library/BaseLib.h>
+#include <Library/DevicePathLib.h>
+#include <Library/DebugLib.h>
+#include <Library/HiiLib.h>
+#include <Library/UefiLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Uefi/UefiMultiPhase.h>
+///
+/// HII specific Vendor Device Path definition.
+///
+typedef struct {
+  VENDOR_DEVICE_PATH             VendorDevicePath;
+  EFI_DEVICE_PATH_PROTOCOL       End;
+} HII_VENDOR_DEVICE_PATH;
+
+extern UINT8 ApuConfigurationUiVfrBin[];
+extern UINT8 ApuConfigurationUiStrings[];
+
+EFI_HANDLE mApuConfigDriverHandle = NULL;
+EFI_HII_HANDLE mApuConfigHiiHandle = NULL;
+
+APU_CONFIGURATION_VARSTORE_DATA mDefaultApuConfig = {

```

Comment:
I like the idea of storing the config in a single variable, but do we know what will happen we add another option here or change their types, and someone updates FW while preserving SMMSTORE?

Reply:
Changing types is rather unacceptable.

If you extend the variable, you should be fine. You can always read an older version of an EFI variable (with smaller size) to C variable (i/e /a struct with a size equal to sizeof). Form should also handle old variable version by applying the old settings to the existing options. And for those non-existing it will apply defaults determined by VFR, i think. If you are curious, you may try adding one option and see what happens.